src/librustc/ty/mod.rs

```diff
@@ -1981,10 +1981,30 @@ impl<'a, 'gcx, 'tcx> AdtDef {
     }
 
     #[inline]
-    pub fn is_non_exhaustive(&self) -> bool {
+    fn is_non_exhaustive(&self) -> bool {
```

Comment:
I would rename this to `has_nonexhaustive_flag` and add a comment about how the meaning of this flag depends a lot on the kind of the ADT. Personally, I find this setup pretty unfortunate: I think we should  be "desugaring" the flag so that it appears on the variant for structs and on the adt only for enums ([as I described here](https://github.com/rust-lang/rust/issues/44109#issuecomment-408127667)). But I guess we can separate that out as a distinct sort of refactoring.